// config/groups.js
var crypto = require('crypto');
var rand = require('csprng');
var mongoose = require('mongoose');
var group = require('config/groupmodel');
var user = require('config/usermodel')


exports.create = function(groupName, groupMemberRequestEmail, groupMemberRequestName, groupAdminEmail,groupAdminName, callback) {

    var token = crypto.createHash('sha512').update(groupName + groupAdminEmail + rand).digest("hex");

    var newgroup = new group({
        token: token,
        groupName: groupName,
        groupRequests: groupMemberRequestEmail,
        groupAdmin: groupAdminEmail,
        groupMembers: groupAdminEmail,
        groupMembersName:groupAdminName,
        groupAdminName:groupAdminName,
        groupRequestsName:groupMemberRequestName

    });

    group.find({
        token: token
    }, function(err, groups) {

        var len = groups.length;

        if (len == 0) {
            newgroup.save(function(err) {

                callback({
                    'response': "Group Sucessfully Created",
                    'res': true
                });

            });

            console.log("okkk"+groupAdminEmail);
            user.find({
                email: groupAdminEmail
            }, function(err, users) {
                users[0].adminGroups.push(token);
                users[0].adminGroupsName.push(groupAdminName);
                users[0].groups.push(token);
                users[0].save(function(err) {
                    console.log("User saved");
                });
            });

            for (var requestMembers in groupMemberRequestEmail) {
                console.log("requestMembers" + groupMemberRequestEmail[requestMembers]);
                user.find({
                    email: groupMemberRequestEmail[requestMembers]
                }, function(err, users) {
                    users[0].groupRequests.push(token);
                    users[0].save(function(err) {
                        console.log("User saved");
                    });
                });
            }
        } else {

            callback({
                'response': "Group already exists",
                'res': false
            });

        }
    });
}


exports.update = function(groupName, groupAdminEmail, userEmail,userName,updateAction, callback) {

    var groupToken = crypto.createHash('sha512').update(groupName + groupAdminEmail + rand).digest("hex");

    switch (updateAction) {
        case "Join Group":
            user.find({
                email: userEmail
            }, function(err, users) {
                var temp = users[0].groupRequests.indexOf(groupToken);
                users[0].groupRequests.splice(temp, 1);
                users[0].groups.push(groupToken);
                users[0].save();
            });

            group.find({
                token: groupToken
            }, function(err, groups) {
            	console.log("error11111"+groupToken);
                var temp = groups[0].groupRequests.indexOf(userEmail);
                groups[0].groupRequests.splice(temp, 1);
                groups[0].groupRequestsName.splice(temp, 1);
                groups[0].groupMembers.push(userEmail);
                groups[0].groupMembersName.push(userName);
                groups[0].save();
            });

            callback({
                'response': "Group Joined Successfully",
                'res': true
            });


            break;
        case "Ignore Group":
        		user.find({
	                email: userEmail
	            }, function(err, users) {
	                var temp = users[0].groupRequests.indexOf(groupToken);
	                users[0].groupRequests.splice(temp, 1);
	                users[0].save();

	            });

        		group.find({
	                token: groupToken
	            }, function(err, groups) {
	                var temp = groups[0].groupRequests.indexOf(userEmail);
	                groups[0].groupRequests.splice(temp, 1);
                    groups[0].groupRequestsName.splice(temp, 1);
	                groups[0].save();
	            });

	            callback({
	                'response': "Group Ignored",
	                'res': true
	            });
        	break;
        case "Leave Group":

            user.find({
                email: userEmail
            }, function(err, users) {
                var temp = users[0].groups.indexOf(groupToken);
                users[0].groups.splice(temp, 1);
                users[0].save();

            });

            group.find({
                token: groupToken
            }, function(err, groups) {
                var temp = groups[0].groupMembers.indexOf(userEmail);
                groups[0].groupMembers.splice(temp, 1);
                groups[0].groupMembersName.splice(temp, 1);
                
                groups[0].save();
            });

            callback({
                'response': "Group Left Successfully",
                'res': true
            });

            break;
        case "Delete Group":
            group.find({
                token: groupToken
            }).remove(function(err) {
                console.log("errrr" + err);
            });


            user.find({
                email: groupAdminEmail
            }, function(err, users) {
                var temp = users[0].groups.indexOf(groupToken);
                users[0].groups.splice(temp, 1);
                //users[0].save();

                var temp1 = users[0].adminGroups.indexOf(groupToken);
                users[0].adminGroups.splice(temp1, 1);
                users[0].save(function(err) {
                    console.log("User saved" + err);
                });

            });

          
                user.find({

                }, function(err, users) {
                	for(var i=0;i<users.length;i++){
                		 var temp = users[i].groups.indexOf(groupToken);
		                    users[i].groups.splice(temp, 1);
		                    var temp1 = users[i].groupRequests.indexOf(groupToken);
		                    users[i].groupRequests.splice(temp1, 1);

		                    users[i].save();
                	}
                   
                });
            
            callback({
                'response': "Group Deleted Successfully",
                'res': true
            });
            break;
    }

}


exports.getGroupDetails = function(groupToken,callback) {

     group.find({
                token: groupToken
            }, function(err, maingroups) {

                
                callback({
                    'response': "Group Name found",
                    'res': true,
                    'groupName': maingroups[0].groupName,
                    'groupAdmin': maingroups[0].groupAdmin,
                    'groupMembers': maingroups[0].groupMembers,
                    'groupAdminName':maingroups[0].groupAdminName,
                    'groupMembersName':maingroups[0].groupMembersName
                });
            });
}
